WordPress/openverse-frontend
# Convert the search store to Pinia

## Fixes


Related to #756 by @sarayourfriend

## Description


This is a bigger PoC for Pinia migration based on the ideas from the [RFC](https://github.com/WordPress/openverse/pull/174). The intention is to see:
- how Pinia stores can be used from Vuex stores and vice versa,
- how to best structure a _setup store_ considering that _setup stores_ give us more flexibility and responsibility for store organization, and
-  to figure out the best testing strategies.

### Converting Vuex mutations
I have removed the simple mutations that existed only because of the Vuex limitations that did not allow mutating state inside actions: `SET_SEARCH_TYPE` and `SET_SEARCH_QUERY`. The effect of these is easy to test within the actions by checking the resulting state.
Larger mutations were converted into Pinia actions.

### Extracting the `filterStore` from the `searchStore`
To make the store more organized, I've extracted the `filterStore` to a separate file. It is private as in its data functions are only used by the `searchStore`, and components do not access it (but they could). My main concern in making it a part of the `search.js` file is the fact that it would make the file very long and hard to read.
We could also just make the `filterStore` a completely separate store, and make components use it directly and not through the `searchStore`. This would simplify the re-exporting of `filterStore` actions and getters by the `searchStore`, but it would cause two `useXStore` lines in the components that use both stores instead of one `useSearchStore` as it is now. I think I actually prefer having `filterStore` as a separate store because re-exporting looks complicated.

#### `searchStore`
-  The `searchStore` only handles the `searchType` and the `query` object, which holds the data about the search term and the filters that are available for it in the form that can be used for the API request.

#### `filterStore`
- Everything related to filters and filter manipulation is  extracted into a `filterStore`. It is an _internal_ store, meaning that it is only used by the `searchStore`. If components need it, they use the `actions` and `getters` of the `searchStore`, which are sometimes simply re-exporting those of the `filterStore`.

#### Working with Vuex stores
- To ensure that we do not have circular imports, `mediaStore` uses the `searchStore` only inside functions.
- `mediaStore`, which is still in Vuex, uses the search store inside functions (getters, actions, mutations).

### Testing
This PR has almost 100% coverage for both filter and search store. It also adds some more e2e tests for the behavior that has been affected by the PR changes.

### Other changes
- some searchStore actions were changed from `async` to `sync` because they do not use network calls, and don't need to be async.
- `mature` filter was refactored to use an object type instead of a boolean to match all other filters.
- `pinia` removed from `package.json` `devDependencies` because it is already listed in `dependencies`.
- removed unused module, action and mutation constants.
- the creation of filters list is moved to `constants` to make the filter store more readable.
- removed a return that's outside of a function in `json-to-pot` (to clean up test output).
- some methods in `search.vue` page were moved into the `setup` function.

## Testing Instructions

Run the app and check that following actions, affected by the changes in this PR, work as expected, both on SSR and on the client:
- filtering (toggling filters, clearing filters, changing applied filters when changing the search type)
- searching various content types.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
